django/contrib/admin/widgets.py

```diff
@@ -307,7 +314,6 @@ def get_context(self, name, value, attrs):
 
         rel_opts = self.rel.model._meta
         info = (rel_opts.app_label, rel_opts.model_name)
-        self.widget.choices = self.choices
```

Comment:
I left the individual commits/reverts as fixups so that you could see - we'll squash them down before merging.

See [here](https://github.com/django/django/pull/16943/commits/41d47bd5ee515056ed70fc6d06c2b0d3d47be456#diff-5a7bf4b2271ba7a2dcce4b0bbef40199985f02cbce6aeebf2111401c2342d57bR899-R900) where I changed it so that, when we set `ChoiceField.choices`, we do re-normalize when assigning to the `.choices` of the widget. This goes back to the original code on the basis that it's [unlikely](https://github.com/django/django/pull/12449#issuecomment-753629874) to be detrimental to performance and I found it hard to ensure that, when setting `._choices` instead, that this also was propagated by `RelatedFieldWidgetWrapper`.

In particular here, the change was to avoid duplicating `self.widget.choices` onto `RelatedFieldWidgetWrapper`, by forwarding to `self.widget` using properties as we do for many other things. That allows for removal of this assignment in `.get_context()` which looks like a hack to ensure that choices of the underlying widget is in sync with the `RelatedFieldWidgetWrapper` widget prior to rendering in case the choices were later altered post-instantiation 🙈

Reply:
I've decided to remove these now at this stage of review - we can bring one or the other of them back back if we need to later.